src/transformers/tokenization_utils_base.py

```diff
@@ -1746,16 +1746,27 @@ def apply_chat_template(
         if tokenizer_kwargs is None:
             tokenizer_kwargs = {}
 
+        using_default_template = False
+
         # First, handle the cases when the model has a dict of multiple templates
         if isinstance(self.chat_template, dict) or (
             self.chat_template is None and isinstance(self.default_chat_template, dict)
         ):
-            template_dict = self.chat_template or self.default_chat_template
+            if self.chat_template is not None:
+                template_dict = self.chat_template
+                using_default_dict = False
+            else:
+                template_dict = self.default_chat_template
+                using_default_dict = True
```

Comment:
Is there ever a case when template_dict = self.default_chat_template and we're not setting using_default_template = True? i.e. do we need this extra flag?

Reply:
Yes - this can happen if:

1. `self.chat_template` is `None`
2. `self.default_chat_template` is a `dict`
3. The user passes a full chat template to the chat template arg, rather than a key to the default dict

Reply:
(I also find thinking through the maze of conditionals here annoying, but rest assured it's all going away the moment I can deprecate default templates!)

Reply:
Gotcha - thanks for being patient with me 😅